File: min-stack/min-stack.java
class MinStack {
    HashMap<Integer, Stack<Integer>> map;
    List<Integer> list;
    int globalMin;
    public MinStack() {
        map = new HashMap<>();
        list = new ArrayList<>();
        globalMin = Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        list.add(val);
        if (!map.containsKey(val)) {
            map.put(val, new Stack<Integer>());
        }
        if (val <= globalMin) {
            map.get(val).push(list.size() - 1);
            globalMin = val;
        } else {
            map.get(val).push(map.get(list.get(list.size() - 2)).peek());
        }
        
    }
    
    public void pop() {
        if (list.isEmpty()) return;
        int val = list.get(list.size() - 1);
        list.remove(list.size() - 1);
        map.get(val).pop();

        if (list.isEmpty()) {
            globalMin = Integer.MAX_VALUE;
        } else {
            globalMin = list.get(map.get(list.get(list.size() - 1)).peek());
        }
    }
    
    public int top() {
        return list.get(list.size() - 1);
    }
    
    public int getMin() {
        return globalMin;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minval = float('inf')

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))
            self.minval = val
        else:
            if val < self.minval:
                self.minval = val
            self.stack.append((val, self.minval))

    def pop(self) -> None:
        _, minimum = self.stack.pop()
        if minimum == self.minval:
            if self.stack:
               self.minval = self.stack[-1][1]
            else:
               self.minval = float('inf')
    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.minval


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()